circuit/src/gadgets/range.rs

```diff
@@ -0,0 +1,471 @@
+	x: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt, const K: usize, const N: usize> LookupRangeCheckChip<F, K, N> {
+	/// Construct a new instance
+	pub fn new(x: AssignedCell<F, F>) -> Self {
+		Self { x }
+	}
+}
+
+impl<F: FieldExt, const K: usize, const N: usize> MockChip<F> for LookupRangeCheckChip<F, K, N> {
+	type Output = AssignedCell<F, F>;
+
+	fn configure(
+		mock_common: &MockCommonConfig, meta: &mut halo2::plonk::ConstraintSystem<F>,
+	) -> Selector {
+		let running_sum_selector = meta.complex_selector();
+
+		let running_sum = mock_common.common.advice[0];
+
+		meta.lookup("running_sum check", |meta| {
+			let running_sum_selector = meta.query_selector(running_sum_selector);
+			let z_cur = meta.query_advice(running_sum, Rotation::cur());
+
+			// In the case of a running sum decomposition, we recover the word from
+			// the difference of the running sums:
+			//    z_i = 2^{K}⋅z_{i + 1} + a_i
+			// => a_i = z_i - 2^{K}⋅z_{i + 1}
+			let z_next = meta.query_advice(running_sum, Rotation::next());
+			let running_sum_word = z_cur.clone() - z_next * F::from(1 << K);
+			let running_sum_lookup = running_sum_selector.clone() * running_sum_word;
+
+			// Combine the running sum and short lookups:
+			vec![(running_sum_lookup, mock_common.table)]
+		});
+
+		running_sum_selector
+	}
+
+	fn synthesize(
+		self, mock_common: &MockCommonConfig, selector: &Selector, mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, halo2::plonk::Error> {
+		layouter.assign_region(
+			|| "range check chip",
+			|region: Region<'_, F>| {
+				let mut ctx = RegionCtx::new(region, 0);
+				let x = ctx.copy_assign(mock_common.common.advice[0], self.x.clone())?;
+
+				// Take first "num_bits" bits of `element`.
+				let num_bits = K * N;
+				let bits = x.value().map(|element| {
+					fe_to_le_bits(element.clone()).into_iter().take(num_bits).collect::<Vec<_>>()
+				});
+
+				// Chunk the "bits" into K-bit words.
+				let words = bits
+					.map(|bits| {
+						bits.chunks_exact(K)
+							.map(|word| F::from(le_bits_to_u64::<K>(&(word.try_into().unwrap()))))
+							.collect::<Vec<_>>()
+					})
+					.transpose_vec(N);
+
+				let mut zs = vec![x.clone()];
+
+				// Assign cumulative sum such that
+				//          z_i = 2^{K}⋅z_{i + 1} + a_i
+				// => z_{i + 1} = (z_i - a_i) / (2^K)
+				//
+				// For `element` = a_0 + 2^10 a_1 + ... + 2^{120} a_{12}}, initialize z_0 =
+				// `element`. If `element` fits in 130 bits, we end up with z_{13} = 0.
+				let mut z = x.clone();
+				let inv_two_pow_k = F::from(1u64 << K).invert().unwrap();
+				for word in words {
+					// Enable q_lookup on this row
+					ctx.enable(selector.clone())?;
+
+					// z_next = (z_cur - m_cur) / 2^K
+					let z_next = z.value().zip(word).map(|(z, word)| (*z - word) * inv_two_pow_k);
+
+					// Assign z_next
+					ctx.next();
+					z = ctx.assign_advice(mock_common.common.advice[0], z_next)?;
+
+					zs.push(z.clone());
```

Comment:
Do we need the whole array, just to get the last value from it?

Reply:
Yes, I think we just need the last cell for zero check.
I removed the array logic from code. https://github.com/eigen-trust/protocol/pull/169/commits/f5cb0f482c9f45b0327720226205386fe8a03613